plugins/woocommerce/src/Blocks/BlockTemplatesRegistry.php

```diff
@@ -0,0 +1,84 @@
+<?php
+namespace Automattic\WooCommerce\Blocks;
+
+use Automattic\WooCommerce\Blocks\Utils\BlockTemplateUtils;
+use Automattic\WooCommerce\Blocks\Templates\AbstractTemplate;
+use Automattic\WooCommerce\Blocks\Templates\AbstractTemplatePart;
+use Automattic\WooCommerce\Blocks\Templates\MiniCartTemplate;
+use Automattic\WooCommerce\Blocks\Templates\CartTemplate;
+use Automattic\WooCommerce\Blocks\Templates\CheckoutTemplate;
+use Automattic\WooCommerce\Blocks\Templates\CheckoutHeaderTemplate;
+use Automattic\WooCommerce\Blocks\Templates\OrderConfirmationTemplate;
+use Automattic\WooCommerce\Blocks\Templates\ProductAttributeTemplate;
+use Automattic\WooCommerce\Blocks\Templates\ProductCatalogTemplate;
+use Automattic\WooCommerce\Blocks\Templates\ProductCategoryTemplate;
+use Automattic\WooCommerce\Blocks\Templates\ProductTagTemplate;
+use Automattic\WooCommerce\Blocks\Templates\ProductSearchResultsTemplate;
+use Automattic\WooCommerce\Blocks\Templates\SingleProductTemplate;
+
+/**
+ * BlockTemplatesRegistry class.
+ *
+ * @internal
+ */
+class BlockTemplatesRegistry {
+
+	/**
+	 * The array of registered templates.
+	 *
+	 * @var AbstractTemplate[]|AbstractTemplatePart[]
+	 */
+	private $templates = array();
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->init();
+	}
+
+	/**
+	 * Initialization method.
+	 */
+	public function init() {
+		if ( BlockTemplateUtils::supports_block_templates( 'wp_template' ) ) {
+			$templates = array(
+				ProductCatalogTemplate::SLUG       => new ProductCatalogTemplate(),
+				ProductCategoryTemplate::SLUG      => new ProductCategoryTemplate(),
+				ProductTagTemplate::SLUG           => new ProductTagTemplate(),
+				ProductAttributeTemplate::SLUG     => new ProductAttributeTemplate(),
+				ProductSearchResultsTemplate::SLUG => new ProductSearchResultsTemplate(),
+				CartTemplate::SLUG                 => new CartTemplate(),
+				CheckoutTemplate::SLUG             => new CheckoutTemplate(),
+				OrderConfirmationTemplate::SLUG    => new OrderConfirmationTemplate(),
+				SingleProductTemplate::SLUG        => new SingleProductTemplate(),
+			);
```

Comment:
I wonder if we should use the DI container for these classes as well. My understanding is that we would need to register them in `Bootstrap.php`, though, which was something [we were trying to avoid](https://github.com/woocommerce/woocommerce/pull/44537#discussion_r1495572238). Thoughts @dinhtungdu?

Reply:
As discussed in DM, I think it makes more sense to have `BlockTemplatesRegistry` to manage all templates. Doing so gives us more control and more safety working with templates.

Templates are managed by `BlockTemplatesRegistry` and still belong to the container.